File: src/timestamp.zig
const std = @import("std");

pub fn hasValidTimestamp(filename: []const u8) bool {
    // Check minimum length (YYYY-MM-DDTHHMMSS = 17 characters)
    if (filename.len < 17) return false;

    // Check all digits and delimiters are in correct positions
    const pattern = "0000-00-00T000000";
    inline for (pattern, 0..) |c, i| {
        switch (c) {
            '0' => {
                if (!std.ascii.isDigit(filename[i])) return false;
            },
            else => {
                if (filename[i] != c) return false;
            },
        }
    }

    // Parse and validate components
    const year = std.fmt.parseInt(u16, filename[0..4], 10) catch return false;
    const month = std.fmt.parseInt(u8, filename[5..7], 10) catch return false;
    const day = std.fmt.parseInt(u8, filename[8..10], 10) catch return false;
    const hour = std.fmt.parseInt(u8, filename[11..13], 10) catch return false;
    const minute = std.fmt.parseInt(u8, filename[13..15], 10) catch return false;
    const second = std.fmt.parseInt(u8, filename[15..17], 10) catch return false;

    // Validate ranges
    if (month < 1 or month > 12) return false;
    if (day < 1 or day > 31) return false;
    if (hour > 23) return false;
    if (minute > 59) return false;
    if (second > 59) return false;

    // Validate days in month
    const days_in_month = switch (month) {
        2 => if (isLeapYear(year)) @as(u8, 29) else @as(u8, 28),
        4, 6, 9, 11 => 30,
        else => 31,
    };
    if (day > days_in_month) return false;

    return true;
}

fn isLeapYear(year: u16) bool {
    return (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0);
}

test "hasValidTimestamp" {
    const testing = std.testing;

    // Valid timestamps
    try testing.expect(hasValidTimestamp("2024-01-16T020000_test.txt"));
    try testing.expect(hasValidTimestamp("2024-02-29T235959_leap.txt"));
    try testing.expect(hasValidTimestamp("2024-12-31T000000"));

    // Invalid timestamps
    try testing.expect(!hasValidTimestamp("2024-13-01T000000_test.txt")); // Invalid month
    try testing.expect(!hasValidTimestamp("2024-04-31T000000_test.txt")); // Invalid day for April
    try testing.expect(!hasValidTimestamp("2023-02-29T000000_test.txt")); // Invalid leap year
    try testing.expect(!hasValidTimestamp("2024-01-16T240000_test.txt")); // Invalid hour
    try testing.expect(!hasValidTimestamp("2024-01-16T006000_test.txt")); // Invalid minute
    try testing.expect(!hasValidTimestamp("not_a_timestamp.txt")); // No timestamp
    try testing.expect(!hasValidTimestamp("2024-01-16_no_time.txt")); // Missing time
    try testing.expect(!hasValidTimestamp("20240116T020000_test.txt")); // Wrong date format
    try testing.expect(!hasValidTimestamp("2024-01-16-020000_test.txt")); // Wrong separator
}

File: src/timestamp.py
"""Timestamp validation for migration filenames."""

import datetime


def has_valid_timestamp(filename: str) -> bool:
    """
    Validate that filename starts with YYYY-MM-DDTHHMMSS format.

    Validation rules:
    - Minimum length of 17 characters
    - Pattern: YYYY-MM-DDTHHMMSS where Y/M/D/H/M/S are digits
    - Hyphens at positions 4 and 7
    - 'T' at position 10
    - Month: 1-12
    - Day: 1-31 (adjusted for month and leap years)
    - Hour: 0-23
    - Minute: 0-59
    - Second: 0-59

    Args:
        filename: The filename to validate

    Returns:
        True if filename has valid timestamp prefix, False otherwise
    """
    # Check minimum length
    if len(filename) < 17:
        return False

    # Check pattern: YYYY-MM-DDTHHMMSS
    # Positions: 0123456789012345678
    pattern = "0000-00-00T000000"

    for i, c in enumerate(pattern):
        if c == "0":
            # Should be a digit
            if not filename[i].isdigit():
                return False
        else:
            # Should match the delimiter
            if filename[i] != c:
                return False

    # Parse and validate components
    try:
        year = int(filename[0:4])
        month = int(filename[5:7])
        day = int(filename[8:10])
        hour = int(filename[11:13])
        minute = int(filename[13:15])
        second = int(filename[15:17])
    except ValueError:
        return False

    # Validate ranges
    if month < 1 or month > 12:
        return False
    if day < 1 or day > 31:
        return False
    if hour > 23:
        return False
    if minute > 59:
        return False
    if second > 59:
        return False

    # Validate days in month using datetime
    try:
        datetime.date(year, month, day)
    except ValueError:
        return False

    return True


def is_leap_year(year: int) -> bool:
    """Check if a year is a leap year."""
    return (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0)
